### PR TITLE
Fix a convergence issues in TP topology caused by incorrect grad_norm.

### PR DESCRIPTION
**Pull request by inkcherry**
Some users are concerned that changes in TP topology during MOE training may potentially cause interference with experiments when noticing similar issues 
https://github.com/microsoft/Megatron-DeepSpeed/issues/151 
https://github.com/microsoft/Megatron-DeepSpeed/pull/176/files

We found a grad_norm calculation error after enabling TP. This error occurs because flattened grad of a params group is used, where the group contains both non-TP and TP parameters. Therefore, it is not possible to use a single attribute to determine whether flattened grad needs to compute the norm. In the current code logic, all params are assumed to be non-TP, resulting in only tp_rank0 grad participating in grad_norm computation. Other tp_rank grads have grad_norm_sum equal to 0. We tested and found that with TP=1 and TP=4, the difference in grad_norm is approximately twice (sqrt(4)). This aligns with the aforementioned issue. This problem should also affect dense models.

Due to the absence of flattening params_group grad in bf16, this problem is avoided. 

We tested the loss curve on the 1.3B model. In cases where TP size increases the inconsistent gap should be larger.

with this change 1.3B with EP=4 TP=4 &1 , fp16,mbs=1,gbs=16
image

without this change  1.3B with EP=4 TP=4&1  ,fp16,mbs=1,gbs=16
image

### CODE REVIEW

**Comment by tjruwase** [comment]
@inkcherry, amazing contribution!!! 

@tohtana, FYI

File: deepspeed/runtime/utils.py [hunk 1]
```diff
@@ -814,25 +824,6 @@ def get_only_unique_item(items):
...
-    """
-    if global_grad_norm is None:
-        global_grad_norm = get_grad_norm(parameters, mpu=mpu)
-    clip_coef = max_norm / (global_grad_norm + eps)
-    if clip_coef < 1:
-        for p in parameters:
-            p.grad.detach().mul_(clip_coef)
-    return global_grad_norm
-
-
```
[the remark this answers is not in the document]
Comment (inkcherry) [reply deepspeed/runtime/utils.py:835]: I modified the ```get_grad_norm``` function, and here the ```clip_gradients``` function is another place where the ```get_grad_norm``` function is called, but clip_gradient function is not used anywhere.

File: deepspeed/runtime/utils.py [hunk 2]
```diff
@@ -415,18 +415,28 @@ def get_grad_norm(parameters, norm_type=2, mpu=None):
...
+
+                # # mask_tensor_ = torch.zeros_like(p, device=p.device, dtype=bool)
+                # #for mask_idx in grad_norm_mask[idx]:
+                # #    mask_tensor_[mask_idx[0]:mask_idx[1]] = True
+                cum_sum_pairs = torch.tensor([1, -1], device=get_accelerator().current_device(),
+                                             dtype=p.dtype).repeat(grad_norm_mask[idx].shape[0], 1)
+                mask_tensor = torch.zeros(p.shape[0] + 1, device=get_accelerator().current_device(), dtype=p.dtype)
+                mask_tensor = mask_tensor.scatter_(0, grad_norm_mask[idx].view(-1),
+                                                   cum_sum_pairs.view(-1)).cumsum(0).bool()[:-1]
+                # assert torch.equal(mask_tensor_, mask_tensor)
```
[the remark this answers is not in the document]
Comment (inkcherry) [reply deepspeed/runtime/utils.py]: deleted

File: deepspeed/runtime/utils.py [hunk 3]
```diff
@@ -415,18 +415,28 @@ def get_grad_norm(parameters, norm_type=2, mpu=None):
...
+                # Use grad_norm_mask to avoid redundant computation of flattened gradient norm
+                # # including, Pipeline parallelism may replicate parameters.
+                # # replicated tensors from tensor model parallelism
+
+                # A loop-free implementation to create a mask tensor based on a range list,
+                # which is logically equivalent to the following implementation.
+
+                # # mask_tensor_ = torch.zeros_like(p, device=p.device, dtype=bool)
+                # #for mask_idx in grad_norm_mask[idx]:
+                # #    mask_tensor_[mask_idx[0]:mask_idx[1]] = True
```
[the remark this answers is not in the document]
Comment (inkcherry) [reply deepspeed/runtime/utils.py]: cleaned up and kept some for readability

**Review by tohtana — APPROVED** [review]
This looks good to me, thank you @inkcherry!

**Review by conglongli — APPROVED** [review]
